Userguide Readme for Workbench (#20)
federated-research/5S-TES-Workbench: README.md

The Workbench is a Python-based library which provides a simple interface for submitting computational tasks to a Task Execution Service (TES) endpoint within a Five Safes federated research environment. It handles configuration validation, authentication, TES task construction, and result retrieval allowing researchers to focus on their analysis rather than the underlying infrastructure.

The typical workflow consists of four steps:

- **Validate:** Provide your infrastructure configuration and credentials.
- **Build:** Choose a task template and supply your analysis parameters.
- **Submit:** Send the task to the TES endpoint.
- **Pool Results:** Fetch output files from MinIO storage once the task completes.

## Pre-Requisites

Before using the Workbench, ensure you have the following:

- **Python 3.13+** installed
- **`five-safes-tes-workbench`** package installed in your environment:
    
    ```bash
    pip install five-safes-tes-workbench
    ```
    
- Access to a running **TES endpoint**
- Access to a **MinIO** instance with an output bucket configured

> - Valid **authentication credentials** either a pre-obtained access token or Keycloak client/user credentials provided by your TRE administrator

Reviewer: ```suggestion
- Valid **authentication credentials**, which are either a pre-obtained access token from Submission Layer UI or Submission API's Keycloak credentials provided by Submission layer's administrator
```

## Validation

Before building or submitting any task, you must validate your configuration and credentials by calling `wb.validate()`.

> The first step will be importing the workbench from the pip package as follows and initialize the client.

Reviewer: lines 35-41 are not related to `Validation`

```
from five_safes_tes_workbench.workbench import Workbench

wb = Workbench()
```

> ### Required Params

Reviewer: agreed with @anwarfg about this section, it seems a bit confusing. I suggest we do something like this:
- A section about required params: we have two types of required params: config and auth 
    - what are config params
    - what are auth params
        - for this, you can do access token or KC client credentials (+ what is the benefits and different between these two modes of auth params) 
- A section about how to pass the required params:
    - you can pass + validate the above params by directly putting them: how? 
    - you can pass + validate the above params by YAML file: how? 

hope it helps? 

you can even put this structure into AI and ask it to refine the sentences as well

    | Parameter | Description |
    | --- | --- |
    | access_token | Pre-obtained access token from the Submission UI |
    | client_id / client_secret | Keycloak client credentials |
    | username / password | Keycloak user credentials |
    | keycloak_url | Keycloak base URL |

#### Option A - Direct Params

The next step is to validate the user using the required parameters. There are two ways to provide your configuration for validation: 

Parse the parameters directly as a keyword arguments. There are two further ways to do it. 

<br>

- **Parsing the Configuration params and Authentication params**

```bash
wb.validate(
		# --- Config Params ---
    project="your-project-name",
    tes_base_url="http://your-tes-endpoint:5034",
    minio_sts_endpoint="http://your-minio-endpoint:9000/sts",
    minio_endpoint="http://your-minio-endpoint:9000",
    minio_output_bucket="your-output-bucket-name",
    tres=["Your-TRE1"],
    
    # --- Auth Params ---
    client_id="your-client-id",
    client_secret="your-client-secret",
    username="your-username",
    password="your-password",
    keycloak_url="http://your-keycloak-endpoint/",
)
```

<br>

> - **Parsing the Configuration params and Access Token**

Reviewer 2: It's not clear that it's
Parsing the Configuration params and Authentication params
OR
Parsing the Configuration params and Access Token.

And it seems strange that that you say earlier you can parse as a keyword arg, then there are two further ways to do it, but you have options A and B.

I see that you mean that you can use args directly OR use a YAML, and in either, you can use the token directly OR use the keycloak credentials. I think this needs to be clearer (I think it's clear in your YAML example because of the comments).

  minio_output_bucket: "your-output-bucket-name"
  tres:
    - "Your-TRE1"
    - "Add more TREs as needed"

  auth:
    # ---- Option 1: Access Token ----
    # (Use Access Token if you have one from the Submission UI.)

    access_token: "your-access-token-here"

    # ---- Option 2: Keycloak Credentials ----
    # (Use Keycloak credentials if you want the Workbench to obtain an access token on your behalf.)

    client_id: "your-keycloak-client-id"
    client_secret: "your-keycloak-client-secret"
    username: "your-keycloak-username"
    password: "your-keycloak-password"
    keycloak_url: "http://localhost:your-keycloak-port/"
```

### **Build TES and Submit**

Once the Workbench has been validated, the next step is to build the TES task. 

The Workbench provides a template-based interface via `wb.build_tes.<template>(...)` which constructs the TES message that will be submitted to the endpoint.

**The available Templates are mentioned below:** 

| Template | Method | Use Case |
| --- | --- | --- |
| Hello World | build_tes.hello_world(...) | Verify the setup is working end-to-end |
| Simple SQL | build_tes.simple_sql(...) | Run a SQL query against a TRE database |

> | Custom | build_tes.custom(...) | Bring your own container image and command |

Reviewer: This should be either at the top or the bottom of this table

| Bunny | build_tes.bunny(...) | Run the Health Informatics Bunny CLI analytics tool |

#### Hello World - wb.build_tes.hello_world( )

The simplest template. Every parameter is optional and can be called with no arguments at all and will fall back to sensible defaults. Use it to verify connectivity before running real analysis.

**Example Implementation**

```bash
# Minimal — no parameters needed
wb.build_tes.hello_world()

wb.submit()
```

#### Simple SQL - wb.build_tes.simple_sql( )

> Executes a SQL query against a SQL database using the federated SQL analysis tool. All the parameters in the simple sql method can be overridden within the template. 

Reviewer: please tell the readers about the default image used in this mode

> | Parameters | Required | Default | Description |

Reviewer: ```suggestion
| Parameters | Required  | Description |
```

| --- | --- | --- | --- |
| name | Yes | — | Task name |
| query | Yes | — | SQL query to execute |

**Example Implementation**

```bash

> _query = (

Reviewer 2: Shouldn't use _query as the variable name, because a leading underscore in python means that it's intended for internal/private use, which is not the usage in this example.

Author: Okay👍🏻

    'WITH user_query AS ('
    'SELECT value_as_number FROM "NottinghamDemo".measurement '
    'WHERE measurement_concept_id = 3000905 '
    'AND value_as_number IS NOT NULL'
    ') SELECT COUNT(*) AS n, SUM(value_as_number) AS total FROM user_query;'
)

wb.build_tes.simple_sql(
    name="Simple SQL Task",
    query=_query,
)

wb.submit()
```

#### Bunny - wb.build_tes.bunny( )

> Runs the Health Informatics Bunny CLI analytics container. The container image is fixed by the template. You provide the command list which is passed directly to the Bunny CLI. 

Reviewer: please provide what image is fixed to use in this mode

All the parameters in the bunny method can be overridden within the template. 

> | Parameter | Required | Default | Description |

Reviewer: ```suggestion
| Parameter | Required  | Description |
```

| --- | --- | --- | --- |
| name | Yes | — | Task name |
| command | Yes | — | Arguments passed to the Bunny CLI |

**Example Implementation**

```bash
wb.build_tes.bunny(
    name="Bunny Task",
    command=[
        "--body-json",
        '{"code":"GENERIC","analysis":"DISTRIBUTION","uuid":"123","collection":"test","owner":"me"}',
        "--output",
        "/outputs/output.json",
        "--no-encode",
    ],
)

wb.submit()
```

#### Custom - wb.build_tes.custom( )

> A fully user-defined task. name, image, and command are required. All other fields fall back to defaults if not provided.

Reviewer 2: Put the var names in backticks, as code.

Author: Okay 👍🏻

All the parameters in the custom method can be overridden within the template. 

> For more information about constructing a custom TES message, view the link [[How to create a TES message](https://ga4gh.github.io/task-execution-schemas/docs/#tag/TaskService/operation/CreateTask)].

Reviewer: ```suggestion
For more information about constructing a custom TES message, view [[the official schema of TES from GA4GH](https://ga4gh.github.io/task-execution-schemas/docs/#tag/TaskService/operation/CreateTask)].
```

**Example Implementation**

```bash
wb.build_tes.custom(
    name="Test Custom mode from Workbench",
    description="custom analysis",
    executors=[
        {
            "image": "ubuntu",
            "command": ["echo", "Hello World"],
            "workdir": "/outputs",
            "stdout": "/outputs/stdout"
        }
    ],
    outputs=[
        {
            "name": "Stdout",
            "description": "Stdout results",
            "url": "s3://",
            "path": "/outputs",
            "type": "DIRECTORY"
        }
    ],
)

wb.submit()
```

> ### **Pool Results**

Reviewer 2: I've got an issue with saying 'pool results' - that means merging or aggregating. I'm pretty sure that's not what you mean, but especially in this product, I hope you can see how that could be misleading.

I think something like 'Collect results' might be more appropriate.

Author: Okay 👍🏻

Once a task has been submitted and completed, you can download the output files from MinIO storage using the fetch methods.

**Note:** The task must have reached a `Completed` status before fetching results.

#### **Output Directory Structure**

Downloaded files are saved in an output/ folder created **next to your notebook**, organised by TRE name and child task ID:

```bash
output/
└── <tre_name>/
    └── <child_task_id>/
        ├── output.json
        └── acro_output_20260501_085731.zip
```

#### Fetch Results - wb.fetch_outputs()

The `wb.fetch_output()` method will pool results from the TRE for the submissions made in the previous steps. This method works based on the parameters defined (see params table below).

| Parameter | Required | Description |
| --- | --- | --- |

> | `tre` | No | TRE name to fetch results for. If omitted, results for **all TREs** in config are downloaded |

Reviewer: ```suggestion
| `tre` | No | TRE name to fetch results for. If omitted, Workbench will attempt to download results for **all TREs** in configs |
```

> | `task_id` | No | Integer task ID. Defaults to the ID from the most recent `wb.submit()` call |

Reviewer: ```suggestion
| `task_id` | No | Integer task ID. Defaults to the task ID from the most recent `wb.submit()` call |
```

> 

Reviewer: missing `output_path` here

- If `task_id` is not provided, the Workbench will automatically use the ID from the most recent `wb.submit()` call in the current session. 

- If no task has been submitted and no `task_id` is passed, the method will raise a `ValueError`.

<br>

**Example Implementation**
```python
# Fetch results for all TREs (uses last submitted task ID)
wb.fetch_outputs()

# Fetch results for a specific TRE
wb.fetch_outputs(tre="Nottingham TRE 01")

# Fetch with an explicit task ID (e.g. from a previous submission)
wb.fetch_outputs(task_id=945)

# Fetch for a specific TRE with an explicit task ID
wb.fetch_outputs(task_id=945, tre="Nottingham TRE 01")
```

> #### **What to Expect**

Reviewer 2: I think this section could do with more detail.

I think it's important to say that it's checking the submission layer to see if the task is concluded (finished, cancelled, whatever), and only then getting it from Minio.

Also, describe the behaviour if some results are not finished but others are, what happens if they are cancelled or failed, what happens if some are ready and others are cancelled/failed?

Thinking about it, maybe the best way to get this across might be a little table, showing the different possible states (in progress, terminated (e.g., failed, cancelled), complete) and the output.

Author: I have updated the docs but still would be great if we could pull the status in the logs so that the user will get updates and they remains in the jupyter notebook.

Reviewer: I think it is a good idea, but take into account the deadline, can we make an issue for this and improve it later? @anwarfg

When fetching results, the Workbench will log each step of the process (eg: see below) and downloaded can be found in the directory.

```bash
INFO | Child task info: 945, status: Completed
INFO | Fetching token from keycloak...
INFO | Keycloak token fetched successfully
INFO | Exchanging bearer token for MinIO credentials via STS
INFO | MinIO client initialised
INFO | Found 2 result object(s) for task 945
INFO | Downloading result object: 945/acro_output_20260501_085731.zip
INFO | Downloaded -> output/Nottingham TRE 01/945/acro_output_20260501_085731.zip
INFO | Downloading result object: 945/output.json
INFO | Downloaded -> output/Nottingham TRE 01/945/output.json
```

> 

Reviewer 2: ```suggestion
If one or more results are not ready, either because they have terminated without releasing results (e.g., failed executor, cancelled task or rejected at egress) or they are still in progress, they will raise a warning and move on to the next task. Results for tasks which have been completed successfully will still be downloaded.
```

# User Guide - 5S-TES-Workbench

This is the user guide for using the Workbench within a Jupyter Notebook to make analysis submissions and retrieve results from the submission layer.

The Workbench is a Python-based library which provides a simple interface for submitting computational tasks to a Task Execution Service (TES) endpoint within a Five Safes federated research environment. It handles configuration validation, authentication, TES task construction, and result retrieval allowing researchers to focus on their analysis rather than the underlying infrastructure.

The typical workflow consists of four steps:

- **Validate:** Provide your infrastructure configuration and credentials.
- **Build:** Choose a task template and supply your analysis parameters.
- **Submit:** Send the task to the TES endpoint.
- **Pool Results:** Fetch output files from MinIO storage once the task completes.

## Pre-Requisites

Before using the Workbench, ensure you have the following:

> - **Python 3.11+** installed

Reviewer: ```suggestion
- **Python 3.13+** installed
```

        "--output",
        "/outputs/output.json",
        "--no-encode",
    ],
)

wb.submit()
```

#### Custom - wb.build_tes.custom( )

A fully user-defined task. name, image, and command are required. All other fields fall back to defaults if not provided.

All the parameters in the custom method can be overridden within the template. 

For more information about constructing a custom TES message, view the link [[How to create a TES message](https://ga4gh.github.io/task-execution-schemas/docs/#tag/TaskService/operation/CreateTask)].

**Example Implementation**

```bash
wb.build_tes.custom(
    name="Test Custom mode from Workbench",
    description="custom analysis",
    executors=[
        {
            "image": "ubuntu",
            "command": ["echo", "Hello World"],
            "workdir": "/outputs",
            "stdout": "/outputs/stdout"
        }
    ],
    outputs=[
        {
            "name": "Stdout",
            "description": "Stdout results",
            "url": "s3://",
            "path": "/outputs",
            "type": "DIRECTORY"
        }
    ],
)

wb.submit()
```

> ### **Pool Results**

Reviewer: please update this according to the newest merged PR. 
So only one command now, and the PR description and the docstring should give you the ideas of how this command works

Author: Okay 👍🏻

# User Guide - 5S-TES-Workbench

This is the user guide for using the Workbench within a Jupyter Notebook to make analysis submissions and retrieve results from the submission layer.

The Workbench is a Python-based library which provides a simple interface for submitting computational tasks to a Task Execution Service (TES) endpoint within a Five Safes federated research environment. It handles configuration validation, authentication, TES task construction, and result retrieval allowing researchers to focus on their analysis rather than the underlying infrastructure.

The typical workflow consists of four steps:

- **Validate:** Provide your infrastructure configuration and credentials.
- **Build:** Choose a task template and supply your analysis parameters.
- **Submit:** Send the task to the TES endpoint.

> - **Pool Results:** Fetch output files from MinIO storage once the task completes.

Reviewer: ```suggestion
- **Fetch Results:** Fetch output files from MinIO storage once the task completes.
```

```
from five_safes_tes_workbench.workbench import Workbench

wb = Workbench()
```

## Validation

Before building or submitting any task, you must validate your configuration and credentials by calling `wb.validate()`.

### Required Params

This section explains the required parameters to validate the user before building the TES message and submitting into the submission layer of the 5S-TES. 

The required parameters are structured into two types. 

#### Config Params

 The configuration parameters are required to establish a connection to the TES endpoint, MinIO storage and define which TREs the task will be submitted to.

    | Parameter | Description |
    | --- | --- |
    | project | Project name |
    | tes_base_url | Base URL of the TES service |
    | minio_sts_endpoint | MinIO STS endpoint URL |
    | minio_endpoint | MinIO endpoint URL |
    | minio_output_bucket | MinIO output bucket name |
    | tres | List of TRE names to target |

<br>

#### Auth Params

The authentication parameters are required to authenticate via ID Provider (Keycloak) and fetch the access token.

You can either provide Keycloak Credentials or Direct Access Key.

**Keycloak Credentials**

>     | Parameter | Description |

Reviewer: these two tables are not showing as tables in the README

<img width="1526" height="484" alt="Image" src="https://github.com/user-attachments/assets/5c6a2028-3a5b-45a0-84ce-bc229a09c627" />

Author: Fixed now.

**The available Templates are mentioned below:** 

| Template | Method | Use Case |
| --- | --- | --- |
| Hello World | build_tes.hello_world(...) | Verify the setup is working end-to-end |
| Simple SQL | build_tes.simple_sql(...) | Run a SQL query against a TRE database |
| Bunny | build_tes.bunny(...) | Run the Health Informatics Bunny CLI analytics tool |
| Custom | build_tes.custom(...) | Bring your own container image and command |

#### Hello World - wb.build_tes.hello_world( )

The simplest template. Every parameter is optional and can be called with no arguments at all and will fall back to sensible defaults. Use it to verify connectivity before running real analysis.

**Example Implementation**

```bash
# Minimal — no parameters needed
wb.build_tes.hello_world()

wb.submit()
```

#### Simple SQL - wb.build_tes.simple_sql( )

Executes a SQL query against a SQL database using the uses default image : `harbor.federated-analytics.ac.uk/5s-tes-analysis-tools/5s-tes-analysis-tools-tre-sqlpg:1.0.0`

| Parameters | Required  | Description |
| --- | --- | --- | --- |
| name | Yes | — | Task name |
| query | Yes | — | SQL query to execute |

> 

Reviewer 2: This table isn't displaying properly (for me).

Reviewer 2: Seems like this might just be an issue for me.

<br>

#### Auth Params

> The authentication parameters are required to authenticate via ID Provider (Keycloak) and fetch the access token.

Reviewer 2: ```suggestion
The authentication parameters are required to authenticate via ID Provider (Keycloak) and fetch the access token (direct access key).
```

You can either provide Keycloak Credentials or Direct Access Key.

**Keycloak Credentials**

    | Parameter | Description |
    | --- | --- |
    | client_id / client_secret | Keycloak client credentials |
    | username / password | Keycloak user credentials |
    | keycloak_url | Keycloak base URL |

    

> **Direct Access Token**

Reviewer 2: Ok, now I think we've called it "access token", "Direct access token" and "direct access key". We should really try to be consistent.

Author: Fixed this one.
